From 20357e1983988353bf1a26c456323054d85d4ab2 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 6 Dec 2018 18:53:05 +0800
Subject: [PATCH] gl-renderer: Workaround dmabuf import/attach for mali

Tested on px3se with libmali-utgard-400-r7p0-r3p0-wayland.so, the
mpp + waylandsink works.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/gl-renderer.c | 37 +++++++++++++++++++++++++++++++------
 1 file changed, 31 insertions(+), 6 deletions(-)

diff --git a/libweston/gl-renderer.c b/libweston/gl-renderer.c
index 2c50d2d..5367b60 100644
--- a/libweston/gl-renderer.c
+++ b/libweston/gl-renderer.c
@@ -260,6 +260,8 @@ struct gl_renderer {
 	PFNEGLCREATESYNCKHRPROC create_sync;
 	PFNEGLDESTROYSYNCKHRPROC destroy_sync;
 	PFNEGLDUPNATIVEFENCEFDANDROIDPROC dup_native_fence_fd;
+
+	int is_mali_egl;
 };
 
 enum timeline_render_point_type {
@@ -1779,6 +1781,14 @@ import_simple_dmabuf(struct gl_renderer *gr,
 	attribs[atti++] = EGL_LINUX_DRM_FOURCC_EXT;
 	attribs[atti++] = attributes->format;
 
+	// Mali needs extra attributes
+	if (gr->is_mali_egl) {
+		attribs[atti++] = EGL_YUV_COLOR_SPACE_HINT_EXT;
+		attribs[atti++] = EGL_ITU_REC601_EXT;
+		attribs[atti++] = EGL_SAMPLE_RANGE_HINT_EXT;
+		attribs[atti++] = EGL_YUV_NARROW_RANGE_EXT;
+  }
+
 	if (attributes->modifier[0] != DRM_FORMAT_MOD_INVALID) {
 		if (!gr->has_dmabuf_import_modifiers)
 			return NULL;
@@ -2113,7 +2123,7 @@ gl_renderer_query_dmabuf_formats(struct weston_compositor *wc,
 
 	if (!gr->has_dmabuf_import_modifiers ||
 	    !gr->query_dmabuf_formats(gr->egl_display, 0, NULL, &num)) {
-		num = gr->has_gl_texture_rg ? ARRAY_LENGTH(fallback_formats) : 2;
+		num = (gr->has_gl_texture_rg || gr->is_mali_egl) ? ARRAY_LENGTH(fallback_formats) : 2;
 		fallback = true;
 	}
 
@@ -2232,7 +2242,7 @@ import_known_dmabuf(struct gl_renderer *gr,
 	return true;
 }
 
-static void
+static int
 gl_renderer_attach_dmabuf(struct weston_surface *surface,
 			  struct weston_buffer *buffer,
 			  struct linux_dmabuf_buffer *dmabuf)
@@ -2246,7 +2256,7 @@ gl_renderer_attach_dmabuf(struct weston_surface *surface,
 	if (!gr->has_dmabuf_import) {
 		linux_dmabuf_buffer_send_server_error(dmabuf,
 				"EGL dmabuf import not supported");
-		return;
+		return -1;
 	}
 
 	buffer->width = dmabuf->attributes.width;
@@ -2284,7 +2294,7 @@ gl_renderer_attach_dmabuf(struct weston_surface *surface,
 
 	if (!import_known_dmabuf(gr, image)) {
 		linux_dmabuf_buffer_send_server_error(dmabuf, "EGL dmabuf import failed");
-		return;
+		return -1;
 	}
 
 	gs->num_images = image->num_images;
@@ -2304,6 +2314,8 @@ gl_renderer_attach_dmabuf(struct weston_surface *surface,
 	gs->height = buffer->height;
 	gs->buffer_type = BUFFER_TYPE_EGL;
 	gs->y_inverted = buffer->y_inverted;
+
+	return 0;
 }
 
 static void
@@ -2339,7 +2351,16 @@ gl_renderer_attach(struct weston_surface *es, struct weston_buffer *buffer)
 	else if (gr->has_bind_display &&
 		 gr->query_buffer(gr->egl_display, (void *)buffer->resource,
 				  EGL_TEXTURE_FORMAT, &format))
+	{
+		// Prefer attaching as dmabuf for mali
+		dmabuf = linux_dmabuf_buffer_get(buffer->resource);
+		if (dmabuf && gr->is_mali_egl) {
+			if (!gl_renderer_attach_dmabuf(es, buffer, dmabuf))
+				return;
+		}
+
 		gl_renderer_attach_egl(es, buffer, format);
+	}
 	else if ((dmabuf = linux_dmabuf_buffer_get(buffer->resource)))
 		gl_renderer_attach_dmabuf(es, buffer, dmabuf);
 	else {
@@ -2815,8 +2836,9 @@ log_extensions(const char *name, const char *extensions)
 }
 
 static void
-log_egl_gl_info(EGLDisplay egldpy)
+log_egl_gl_info(struct gl_renderer *gr)
 {
+	EGLDisplay egldpy = gr->egl_display;
 	const char *str;
 
 	str = eglQueryString(egldpy, EGL_VERSION);
@@ -2843,6 +2865,9 @@ log_egl_gl_info(EGLDisplay egldpy)
 	str = (char *)glGetString(GL_RENDERER);
 	weston_log("GL renderer: %s\n", str ? str : "(null)");
 
+	if (strstr(str, "mali") || strstr(str, "Mali"))
+		gr->is_mali_egl = 1;
+
 	str = (char *)glGetString(GL_EXTENSIONS);
 	log_extensions("GL extensions", str ? str : "(null)");
 }
@@ -3713,7 +3738,7 @@ gl_renderer_setup(struct weston_compositor *ec, EGLSurface egl_surface)
 		gr->gl_version = GR_GL_VERSION(2, 0);
 	}
 
-	log_egl_gl_info(gr->egl_display);
+	log_egl_gl_info(gr);
 
 	gr->image_target_texture_2d =
 		(void *) eglGetProcAddress("glEGLImageTargetTexture2DOES");
-- 
2.11.0

